<img src="img/Marca-ITBA-Color-ALTA.png" width="250">

# Programación para el Análisis de Datos

## Clase 4 parte 1 - Práctica

In [2]:
import numpy as np
import pandas as pd
print(pd.__version__)

df = pd.read_csv('data/phone_data.csv', index_col=0)
df.head()

2.2.2


date  duration  item    month   network network_type
index                                                                
0      15/10/14 06:58    34.429  data  2014-11      data         data
1      15/10/14 06:58    13.000  call  2014-11  Vodafone       mobile
2      15/10/14 14:46    23.000  call  2014-11    Meteor       mobile
3      15/10/14 14:48     4.000  call  2014-11     Tesco       mobile
4      15/10/14 17:27     4.000  call  2014-11     Tesco       mobile

### Ejercicio 1

Particionar el `Dataframe` por `network_type`. Mostrar por pantalla unicamente el grupo `'network_type'='data'`

**Tips de funciones a utilizar:**
  - groupby
  - get_group

In [4]:
### Completa con tu código aquí
grouped_network_type = df.groupby('network_type')

grouped_network_type.get_group('data')

date  duration  item    month network network_type
index                                                              
0      15/10/14 06:58    34.429  data  2014-11    data         data
6      16/10/14 06:58    34.429  data  2014-11    data         data
13     17/10/14 06:58    34.429  data  2014-11    data         data
26     18/10/14 06:58    34.429  data  2014-11    data         data
39     19/10/14 06:58    34.429  data  2014-11    data         data
...               ...       ...   ...      ...     ...          ...
821    09/03/15 06:58    34.429  data  2015-03    data         data
822    10/03/15 06:58    34.429  data  2015-03    data         data
823    11/03/15 06:58    34.429  data  2015-03    data         data
824    12/03/15 06:58    34.429  data  2015-03    data         data
827    13/03/15 06:58    34.429  data  2015-03    data         data

[150 rows x 6 columns]

### Ejercicio 2

Obtener el registro con duración mas larga para el grupo `'landline'`

**Tips de funciones a utilizar:**
  - groupby
  - get_group
  - sort_values

In [22]:
#output_esperado =  10528.0


### Completa con tu código aquí
grouped_network_type.groups.keys()
grouped_network_type.get_group('landline')[['duration']].max()['duration']

# grouped_network_type.get_group('landline').sort_values(by='duration', ascending=False).head(1)['duration'].values[0]
###


10528.0

### Ejercicio 3

Obtener la duración media de los eventos agrupados por `'month'` y`'item'`. 

Obtener, además, la última fecha de evento.

**Tips de funciones a utilizar:**
  - groupby
  - aggregate / agg
  - mean
  - last
  
  
output_esperado:

|                     |   duration | date           |
|:--------------------|-----------:|:---------------|
| ('2014-11', 'call') |    238.757 | 12/11/14 19:01 |
| ('2014-11', 'data') |     34.429 | 12/11/14 06:58 |
| ('2014-11', 'sms')  |      1     | 13/11/14 22:31 |
| ('2014-12', 'call') |    171.658 | 14/12/14 19:54 |
| ('2014-12', 'data') |     34.429 | 12/12/14 06:58 |
| ('2014-12', 'sms')  |      1     | 07/12/14 23:22 |
| ('2015-01', 'call') |    193.977 | 14/01/15 20:47 |
| ('2015-01', 'data') |     34.429 | 12/01/15 06:58 |
| ('2015-01', 'sms')  |      1     | 14/01/15 23:36 |
| ('2015-02', 'call') |    215.164 | 09/02/15 17:54 |
| ('2015-02', 'data') |     34.429 | 12/02/15 06:58 |
| ('2015-02', 'sms')  |      1     | 10/02/15 21:40 |
| ('2015-03', 'call') |    462.277 | 04/03/15 12:29 |
| ('2015-03', 'data') |     34.429 | 13/03/15 06:58 |
| ('2015-03', 'sms')  |      1     | 14/03/15 00:16 |

In [24]:
### Completa con tu código aquí
grouped_month_item = df.groupby(['month', 'item'])
grouped_month_item.aggregate(
    {
         'duration': 'mean',
         'date': 'last'
    }
)
###

duration            date
month   item                            
2014-11 call  238.757009  12/11/14 19:01
        data   34.429000  12/11/14 06:58
        sms     1.000000  13/11/14 22:31
2014-12 call  171.658228  14/12/14 19:54
        data   34.429000  12/12/14 06:58
        sms     1.000000  07/12/14 23:22
2015-01 call  193.977273  14/01/15 20:47
        data   34.429000  12/01/15 06:58
        sms     1.000000  14/01/15 23:36
2015-02 call  215.164179  09/02/15 17:54
        data   34.429000  12/02/15 06:58
        sms     1.000000  10/02/15 21:40
2015-03 call  462.276596  04/03/15 12:29
        data   34.429000  13/03/15 06:58
        sms     1.000000  14/03/15 00:16

### Ejercicio 4

Obtener los primeros `n=2` elementos de cada grupo considerando la columna `network_type` como clave para agrupar.

**Tips de funciones a utilizar:**
  - groupby
  - head
  
Output esperado:

|   index | date           |   duration | item   | month   | network   | network_type   |
|--------:|:---------------|-----------:|:-------|:--------|:----------|:---------------|
|       0 | 15/10/14 06:58 |     34.429 | data   | 2014-11 | data      | data           |
|       6 | 16/10/14 06:58 |     34.429 | data   | 2014-11 | data      | data           |
|      49 | 20/10/14 13:56 |      6     | call   | 2014-11 | landline  | landline       |
|      59 | 23/10/14 08:34 |   1940     | call   | 2014-11 | landline  | landline       |
|       1 | 15/10/14 06:58 |     13     | call   | 2014-11 | Vodafone  | mobile         |
|       2 | 15/10/14 14:46 |     23     | call   | 2014-11 | Meteor    | mobile         |
|     121 | 03/11/14 08:40 |      1     | sms    | 2014-11 | special   | special        |
|     596 | 15/01/15 12:23 |      1     | sms    | 2015-02 | special   | special        |
|       9 | 16/10/14 15:30 |     19     | call   | 2014-11 | voicemail | voicemail      |
|      32 | 18/10/14 14:01 |     96     | call   | 2014-11 | voicemail | voicemail      |
|     360 | 06/12/14 18:27 |      1     | sms    | 2014-12 | world     | world          |
|     361 | 06/12/14 18:28 |      1     | sms    | 2014-12 | world     | world          |

In [55]:
### Tu código aquí:
grouped_by_network_type = df.groupby('network_type')
grouped_by_network_type.head(2)
###


date  duration  item    month    network network_type
index                                                                 
0      15/10/14 06:58    34.429  data  2014-11       data         data
1      15/10/14 06:58    13.000  call  2014-11   Vodafone       mobile
2      15/10/14 14:46    23.000  call  2014-11     Meteor       mobile
6      16/10/14 06:58    34.429  data  2014-11       data         data
9      16/10/14 15:30    19.000  call  2014-11  voicemail    voicemail
32     18/10/14 14:01    96.000  call  2014-11  voicemail    voicemail
49     20/10/14 13:56     6.000  call  2014-11   landline     landline
59     23/10/14 08:34  1940.000  call  2014-11   landline     landline
121    03/11/14 08:40     1.000   sms  2014-11    special      special
360    06/12/14 18:27     1.000   sms  2014-12      world        world
361    06/12/14 18:28     1.000   sms  2014-12      world        world
596    15/01/15 12:23     1.000   sms  2015-02    special      special

### Ejercicio 5

Crear una función que descarte los ultimos `n=200` registros de cada mes de facturación.

**Tips de funciones a utilizar:**
  - groupby
  - tail
  - index
  
  
index_esperado = 
[0,  1,  2,  3,  4,  5,  6,  7,  8,  9,  10,  11,  12,  13,  14,  15,  16,  17,  
18,  19,  20,  21,  22,  23,  24,  25,  26,  27,  28,  29,  381,  386,  389,  390,  391]

In [59]:
### Completa con tu código aquí
group_by_month = df.groupby('month')
group_by_month.groups.keys()

index_list = []
for month in group_by_month.groups.keys():
    partial_index = group_by_month.get_group(month).head(len(group_by_month.get_group(month))-200).index
    print(partial_index)
    index_list.extend(partial_index)

index_list
###

Index([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29],
      dtype='int64', name='index')
Index([228, 231, 232, 233, 234, 235, 236, 237, 238, 239,
       ...
       334, 335, 336, 337, 338, 339, 340, 341, 342, 343],
      dtype='int64', name='index', length=114)
Index([381, 386, 389, 390, 391], dtype='int64', name='index')
Index([577, 586, 594, 595, 596, 597, 598, 599, 600, 601, 602, 603, 604, 605,
       606, 607, 608, 609, 610, 611, 612, 613, 614, 615, 616, 617, 618, 619,
       620, 621, 622, 623, 624, 625, 626, 627, 628, 629, 630, 631, 632, 633,
       634, 635, 636, 637, 638, 639, 640, 641, 642, 643, 644, 645, 646, 647,
       648, 649, 650, 651, 652, 653, 654, 655, 656, 657, 658, 659, 660, 661,
       662, 663, 664, 665],
      dtype='int64', name='index')
Index([729, 730], dtype='int64', name='index')


[0,
 1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 228,
 231,
 232,
 233,
 234,
 235,
 236,
 237,
 238,
 239,
 240,
 241,
 242,
 243,
 244,
 245,
 246,
 247,
 248,
 249,
 250,
 251,
 252,
 253,
 254,
 255,
 256,
 257,
 258,
 259,
 260,
 261,
 262,
 263,
 264,
 265,
 266,
 267,
 268,
 269,
 270,
 271,
 272,
 273,
 274,
 275,
 276,
 277,
 278,
 279,
 280,
 281,
 282,
 283,
 284,
 285,
 286,
 287,
 288,
 289,
 290,
 291,
 292,
 293,
 294,
 295,
 296,
 297,
 298,
 299,
 300,
 301,
 302,
 303,
 304,
 305,
 306,
 307,
 308,
 309,
 310,
 311,
 312,
 313,
 314,
 315,
 316,
 317,
 318,
 319,
 320,
 321,
 322,
 323,
 324,
 325,
 326,
 327,
 328,
 329,
 330,
 331,
 332,
 333,
 334,
 335,
 336,
 337,
 338,
 339,
 340,
 341,
 342,
 343,
 381,
 386,
 389,
 390,
 391,
 577,
 586,
 594,
 595,
 596,
 597,
 598,
 599,
 600,
 601,
 602,
 603,
 604,
 605,
 606,
 607,
 608,
 609,
 610,
 611,
 612,
 613,
 614,
 615,
 6

### Ejercicio 6

* Generar una columna que contenga la hora de las llamadas

* Discretizar la hora de modo tal que se mapeen los rangos horarios a las siguientes categorías:

    - 0-5am - night
    - 5-12 - morning
    - 12-20 - afternoon
    - 20 - 0 evening
    
    
* Hacer una tabla pivot que muestre la duración media de las llamadas en función de la franja horaria y del mes

* Hacer una tabla pivot que muestre el número de llamadas en función de la franja horaria y del mes


In [72]:
### Completa con tu código aquí
df['date'] = pd.to_datetime(df['date'], format='%d/%m/%y %H:%M')
df['hour'] = df['date'].dt.hour
df['hour_category'] = pd.cut(df['hour'], bins=[0, 5, 12, 20, 24], labels=['night', 'morning', 'afternoon', 'evening'])
df.head()
pv1 = df.groupby(['month', 'hour_category'])[['duration']].mean()
pv1.head(20)

pv2 = df.groupby(['month', 'hour_category'])[['duration']].count()
pv2.head(20)










/var/folders/3t/3xp_xcgx09jd49gz6f6db_gr0000gn/T/ipykernel_21497/2634665354.py:6: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  pv1 = df.groupby(['month', 'hour_category'])[['duration']].mean()
/var/folders/3t/3xp_xcgx09jd49gz6f6db_gr0000gn/T/ipykernel_21497/2634665354.py:9: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  pv2 = df.groupby(['month', 'hour_category'])[['duration']].count()


duration
month   hour_category          
2014-11 night                 7
        morning              78
        afternoon           121
        evening              24
2014-12 night                 5
        morning              76
        afternoon            69
        evening               6
2015-01 night                 0
        morning              84
        afternoon            87
        evening              31
2015-02 night                 0
        morning              57
        afternoon            71
        evening               7
2015-03 night                 0
        morning              48
        afternoon            41
        evening               8